File: backend/izin_talep_sistemi/src/main/java/com/kutalmis/izin_talep_sistemi/entity/LeaveBalance.java
package com.kutalmis.izin_talep_sistemi.entity;

import jakarta.persistence.*;

@Entity
@Table(name = "leave_balances", uniqueConstraints = @UniqueConstraint(columnNames = { "user_id", "leave_type_id",
        "year" }))
public class LeaveBalance {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToOne
    @JoinColumn(name = "leave_type_id", nullable = false)
    private LeaveType leaveType;

    @Column(nullable = false)
    private Integer year;

    @Column(nullable = false)
    private Integer totalDays;

    @Column(nullable = false)
    private Integer usedDays = 0;

    @Column(nullable = false)
    private Integer reservedDays = 0;

    @Transient
    public int getAvailableDays() {
        return totalDays - usedDays - reservedDays;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public User getUser() {
        return user;
    }

    public void setUser(User user) {
        this.user = user;
    }

    public LeaveType getLeaveType() {
        return leaveType;
    }

    public void setLeaveType(LeaveType leaveType) {
        this.leaveType = leaveType;
    }

    public Integer getYear() {
        return year;
    }

    public void setYear(Integer year) {
        this.year = year;
    }

    public Integer getTotalDays() {
        return totalDays;
    }

    public void setTotalDays(Integer totalDays) {
        this.totalDays = totalDays;
    }

    public Integer getUsedDays() {
        return usedDays;
    }

    public void setUsedDays(Integer usedDays) {
        this.usedDays = usedDays;
    }

    public Integer getReservedDays() {
        return reservedDays;
    }

    public void setReservedDays(Integer reservedDays) {
        this.reservedDays = reservedDays;
    }

}

File: flutter/izin_talep_sistemi/lib/models/leave_balance.dart
class LeaveBalance {
  final int id;
  final int userId;
  final String userName;
  final String leaveTypeName;
  final int year;
  final int totalDays;
  final int usedDays;
  final int reservedDays;
  final int availableDays;

  LeaveBalance({
    required this.id,
    required this.userId,
    required this.userName,
    required this.leaveTypeName,
    required this.year,
    required this.totalDays,
    required this.usedDays,
    required this.reservedDays,
    required this.availableDays,
  });

  factory LeaveBalance.fromJson(Map<String, dynamic> json) {
    return LeaveBalance(
      id: json['id'],
      userId: json['userId'],
      userName: json['userName'],
      leaveTypeName: json["leaveTypeName"],
      year: json['year'],
      totalDays: json['totalDays'],
      usedDays: json['usedDays'],
      reservedDays: json['reservedDays'],
      availableDays: json['availableDays'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'userId': userId,
      'userName': userName,
      'leaveTypeName': leaveTypeName,
      'year': year,
      'totalDays': totalDays,
      'usedDays': usedDays,
      'reservedDays': reservedDays,
      'availableDays': availableDays,
    };
  }
}
